### DISCUSSION TITLE
what is the relation and differences between metals and sbt

### DISCUSSION THREAD

**Discussion by Tasselmi** [scalameta/metals#3286] [General]
I am now switch from idea to vscode. Just curious about the relation and diffs between the two. Thanks.

**Comment by ckipp01** [comment]
Hey @Tasselmi, when thinking about sbt vs your IDE not much changes if you're using Metals or IntelliJ. sbt is the build tool of your project meaning that your build definition is defined there. Just with sbt and your code you can compile it, run tests, and all sorts of other things like publishing, license checks, etc.

IDE's like IntelliJ and Metals (not technically an IDE, but a language server, but the distinction isn't important here) need to know _about_ your build definition in order to know what version of Scala you're using, what dependencies you're using, and basically know the general structure of your project. Many of the language features you expect from your editor like "goto definition", "find references", or "hovers" are all fully powered by your IDE with the information that your build definition has given them.


So to summarize, you can sort of think about it like this

### sbt

- holds your build definition
- allows you to run, compile, test your code with sbt shell
- allows to you run all sorts of other tasks like publishing, license checks, doc checks, etc

### your IDE

- _uses_ your build definition to know about your projects
- provides all the navigation and other IDE features you'd expect in your editor

Hope this helps. If you want a much longer explanation about how Metals all works together with various tools you can check out [this video](https://www.youtube.com/watch?v=fpzN_vTBy18&t).

**Reply by Tasselmi** [reply]
OK, thanks so much. I think I got what you said. Vscode is like a text editor, and metals enpowers vscode as an IDE, this IDE uses different compiling tools , SBT is just one of the tools.
